### PR TITLE
Add `CQM.from_bqm`, `CQM.from_dqm` and `CQM.from_qm`

### PR DESCRIPTION
**Pull request by arcondello**
Closes https://github.com/dwavesystems/dimod/issues/866

### CODE REVIEW

File: dimod/constrained.py [hunk 1]
```diff
@@ -442,6 +444,60 @@ def add_variable(self, v: Variable, vartype: VartypeLike):
...
+                    )
+
+        objective.offset = dqm.offset
+
+        cqm.set_objective(objective)
+
+        return cqm
+
+    @classmethod
+    def from_quadratic_model(cls, qm: BinaryQuadraticModel) -> 'ConstrainedQuadraticModel':
```
Comment (hhtong) [inline dimod/constrained.py:495]:
```suggestion
    def from_quadratic_model(cls, qm: QuadraticModel) -> 'ConstrainedQuadraticModel':
```

File: dimod/discrete/discrete_quadratic_model.py [hunk 2]
```diff
@@ -537,6 +537,10 @@ def from_numpy_vectors(cls, case_starts, linear_biases, quadratic,
 
         return obj
 
+    def get_cases(self, v):
```
Comment (hhtong) [inline dimod/discrete/discrete_quadratic_model.py:540]: do we need to add this method to dqm.rst?

**Review by hhtong — APPROVED** [review]

File: dimod/discrete/discrete_quadratic_model.py (hunk 2 above)
Comment (arcondello) [reply dimod/discrete/discrete_quadratic_model.py:540]: Also the changelog I suppose

**Comment by hhtong** [comment]
Also, is the plan to later add `from_bqm`, `from_dqm` and `from_qm` as aliases?

**Comment by arcondello** [comment]
> Also, is the plan to later add `from_bqm`, `from_dqm` and `from_qm` as aliases?

I definitely considered it. It felt like it was adding clutter, but I am definitely open to the idea. Do you think they're helpful?

**Comment by hhtong** [comment]
> > Also, is the plan to later add `from_bqm`, `from_dqm` and `from_qm` as aliases?
> 
> I definitely considered it. It felt like it was adding clutter, but I am definitely open to the idea. Do you think they're helpful?

I mean personally I prefer the shorthand. I guess it depends on how much we think these functions will be used? They seem pretty helpful to me but I agree that it would be a bit unnecessary if not used often.

**Comment by randomir** [comment]
I would prefer shorthands as well. Maybe even adding something like `.from_dimod_model` that would do a single dispatch...

**Comment by arcondello** [comment]
Ok, reworked, now there is `CQM.from_discrete_quadratic_model` with alias `CQM.from_dqm` and  `CQM.from_quadratic_model` with aliases `CQM.from_qm` and `CQM.from_bqm`.

File: dimod/constrained.py [hunk 3]
```diff
@@ -442,6 +444,80 @@ def add_variable(self, v: Variable, vartype: VartypeLike):
...
             return self.variables._append(vartype, v)
 
+    @classmethod
+    def from_bqm(cls, bqm: BinaryQuadraticModel) -> 'ConstrainedQuadraticModel':
+        """Alias for :meth:`from_quadratic_model`."""
+        return cls.from_quadratic_model(bqm)
+
+    @classmethod
+    def from_discrete_quadratic_model(cls, dqm: DiscreteQuadraticModel, *,
+                                      relabel_func: Callable[[Variable, int], Variable] = lambda v, c: (v, c),
```
Comment (randomir) [inline dimod/constrained.py:454]: Might be a rare case, with no practical/obvious abuse scenario that comes to mind, but function objects technically are mutable (so shouldn't be used for default argument values).

File: dimod/constrained.py [hunk 4]
```diff
@@ -442,6 +444,80 @@ def add_variable(self, v: Variable, vartype: VartypeLike):
         else:
             return self.variables._append(vartype, v)
 
+    @classmethod
+    def from_bqm(cls, bqm: BinaryQuadraticModel) -> 'ConstrainedQuadraticModel':
```
Comment (randomir) [inline dimod/constrained.py:448]: Do we want to add `from_binary_quadratic_model` for completeness? :shrug:

**Review by randomir — APPROVED** [review]
LGTM!

File: dimod/constrained.py (hunk 3 above)
Comment (arcondello) [reply dimod/constrained.py:454]: Sure, technically. But as you say, I really cannot think of an abuse scenario is this is clear and simple.

File: dimod/constrained.py (hunk 4 above)
Comment (arcondello) [reply dimod/constrained.py:448]: We'll see if anyone requests it.

**Pull request merged**